### PR TITLE
[GraphBolt] Modify Feature class.

### PR DESCRIPTION
**Pull request by yxy235**
## Description

Added tensor dimension detection. Prohibit one-dimensional tensor as feature, and modify the **example** and **update api** in the class on this basis.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/gpu_cached_feature.py [hunk 1]
```diff
@@ -19,24 +19,28 @@ def __init__(self, fallback_feature: Feature, cache_size: int):
         ----------
         fallback_feature : Feature
             The fallback feature.
+            Note that its dimension should be greater than 1.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature.py:22]: Not worth to add it here,  this restriction is related to gpu cached feature.

File: python/dgl/graphbolt/impl/gpu_cached_feature.py [hunk 2]
```diff
@@ -19,24 +19,28 @@ def __init__(self, fallback_feature: Feature, cache_size: int):
...
         >>> feature.read()
-        tensor([0, 1, 2, 3, 4, 5, 6, 7])
-        >>> feature.read(torch.tensor([0, 1, 2]))
-        tensor([0, 1, 2])
-        >>> feature.update(torch.ones(3, dtype=torch.long),
-        ... torch.tensor([0, 1, 2]))
-        >>> feature.read(torch.tensor([0, 1, 2, 3]))
-        tensor([1, 1, 1, 3])
+        tensor([[0, 1, 2, 3, 4],
+            [0, 1, 2, 3, 4]], device='cuda:0')
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature.py:36]:
```suggestion
        tensor([[0, 1, 2, 3, 4],
                [0, 1, 2, 3, 4]], device='cuda:0')
```

File: python/dgl/graphbolt/impl/gpu_cached_feature.py [hunk 3]
```diff
@@ -102,9 +106,10 @@ def update(self, value: torch.Tensor, ids: torch.Tensor = None):
                 value[:size].to("cuda").reshape(self.flat_shape),
             )
         else:
-            assert ids.shape[0] == value.shape[0], (
-                f"ids and value must have the same length, "
-                f"but got {ids.shape[0]} and {value.shape[0]}."
+            assert self._fallback_feature.read(ids).shape == value.shape, (
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/gpu_cached_feature.py:109]: This is a pretty heavy assert.
We should just check the `ids.shape[0] == value.shape[0]` and check the value feature size matches the _fallback_feature's size.

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 4]
```diff
@@ -23,37 +23,48 @@ def __init__(self, torch_feature: torch.Tensor):
...
+                [1, 1, 1, 1, 1]])
 
         >>> import numpy as np
         >>> arr = np.arange(0, 5)
         >>> np.save("/tmp/arr.npy", arr)
         >>> torch_feat = torch.as_tensor(np.load("/tmp/arr.npy",
-        ...         mmap_mode="r+"))
-        >>> feature_store = TorchBasedFeature(torch_feat)
+        ...                                      mmap_mode="r+")
+        ...                              for _ in range(2)])
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:50]: why do we need this line?

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 5]
```diff
@@ -23,37 +23,48 @@ def __init__(self, torch_feature: torch.Tensor):
...
+        >>> feature_store.read(torch.tensor([0]))
+        tensor([[0, 1, 2, 3, 4]])
         """
         super().__init__()
         assert isinstance(torch_feature, torch.Tensor), (
             f"torch_feature in TorchBasedFeature must be torch.Tensor, "
             f"but got {type(torch_feature)}."
         )
+        assert len(torch_feature.shape) > 1, (
+            f"dimension of torch_feature in TorchBasedFeature"
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:64]: nit: make full use of the 80 chars before change to new line.

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 6]
```diff
@@ -99,9 +110,9 @@ def update(self, value: torch.Tensor, ids: torch.Tensor = None):
             )
             self._tensor[:] = value
         else:
-            assert ids.shape[0] == value.shape[0], (
-                f"ids and value must have the same length, "
-                f"but got {ids.shape[0]} and {value.shape[0]}."
+            assert self._tensor[ids].shape == value.shape, (
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:113]: ditto.

File: python/dgl/graphbolt/impl/gpu_cached_feature.py [hunk 7]
```diff
@@ -19,24 +19,28 @@ def __init__(self, fallback_feature: Feature, cache_size: int):
...
+            Note that its dimension should be greater than 1.
         cache_size : int
             The capacity of the GPU cache, the number of features to store.
 
         Examples
         --------
         >>> import torch
-        >>> torch_feat = torch.arange(0, 8)
+        >>> from dgl import graphbolt as gb
+        >>> torch_feat = torch.tensor([range(0, 5), range(0, 5)]).to("cuda")
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/gpu_cached_feature.py:30]: or `torch.arange(10).reshape(2, -1)`?

File: python/dgl/graphbolt/impl/gpu_cached_feature.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/gpu_cached_feature.py:109]: and `assert ids.dim() == 1`?

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 8]
```diff
@@ -23,37 +23,48 @@ def __init__(self, torch_feature: torch.Tensor):
...
+                [0, 1, 2, 3, 4]])
+        >>> feature_store.read(torch.tensor([0]))
+        tensor([[0, 1, 2, 3, 4]])
         """
         super().__init__()
         assert isinstance(torch_feature, torch.Tensor), (
             f"torch_feature in TorchBasedFeature must be torch.Tensor, "
             f"but got {type(torch_feature)}."
         )
+        assert len(torch_feature.shape) > 1, (
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:63]: `assert torch_feature.dim() > 1`?

File: python/dgl/graphbolt/impl/torch_based_feature_store.py (hunk 4 above)
Comment (yxy235) [reply python/dgl/graphbolt/impl/torch_based_feature_store.py:50]: > why do we need this line?

I have forgotten a '[' in L48 before np. I used L50 to create a 2*5 tensor. It can be removed.

File: python/dgl/graphbolt/impl/torch_based_feature_store.py (hunk 6 above)
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:113]: we do the sanity check in `GPUCachedFeature` or `TorchBasedFeature`, is it possible to do such checks in a more unified way?

File: tests/python/pytorch/graphbolt/impl/test_basic_feature_store.py [hunk 9]
```diff
@@ -5,8 +5,8 @@
 
 
 def test_basic_feature_store_homo():
-    a = torch.tensor([3, 2, 1])
-    b = torch.tensor([2, 5, 3])
+    a = torch.tensor([[3], [2], [1]])
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_basic_feature_store.py:8]: is 1-D tensor scenario covered? assertion failure is expected.

File: tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py [hunk 10]
```diff
@@ -12,7 +12,7 @@
     reason="GPUCachedFeature requires a GPU.",
 )
 def test_gpu_cached_feature():
-    a = torch.tensor([1, 2, 3]).to("cuda").float()
+    a = torch.tensor([[1], [2], [3]]).to("cuda").float()
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_gpu_cached_feature.py:15]: I think `a` is totally covered by `b` which is multiple dimension tensor. seems we could remove `a`. Or `a` is a mimic of real scenario such as feature data, label, or something else?

File: python/dgl/graphbolt/impl/torch_based_feature_store.py (hunk 6 above)
Comment (yxy235) [reply python/dgl/graphbolt/impl/torch_based_feature_store.py:113]: > we do the sanity check in `GPUCachedFeature` or `TorchBasedFeature`, is it possible to do such checks in a more unified way?

This change will be withdrawn, so that the checks will be in a unified way. I suggest that this check can be removed in `GPUCachedFeature` for it will call the `update API` in `TorchBasedFeature`? Also, a check with feature size will be added once the modification of feature class have been done.

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 11]
```diff
@@ -23,37 +23,45 @@ def __init__(self, torch_feature: torch.Tensor):
...
+        tensor([[0, 1, 2, 3, 4]])
         """
         super().__init__()
         assert isinstance(torch_feature, torch.Tensor), (
             f"torch_feature in TorchBasedFeature must be torch.Tensor, "
             f"but got {type(torch_feature)}."
         )
+        assert torch_feature.dim() > 1, (
+            f"dimension of torch_feature in TorchBasedFeature must be greater"
+            f"than 1, but got {len(torch_feature.shape)} dimension."
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:63]: use `.dim()` to keep consistent?

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 12]
```diff
@@ -23,37 +23,45 @@ def __init__(self, torch_feature: torch.Tensor):
...
+        tensor([[0, 1, 2, 3, 4],
+                [1, 1, 1, 1, 1]])
 
         >>> import numpy as np
         >>> arr = np.arange(0, 5)
         >>> np.save("/tmp/arr.npy", arr)
-        >>> torch_feat = torch.as_tensor(np.load("/tmp/arr.npy",
-        ...         mmap_mode="r+"))
-        >>> feature_store = TorchBasedFeature(torch_feat)
+        >>> torch_feat = torch.as_tensor([np.load("/tmp/arr.npy",
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:48]: converting 1-D array to multi-dim via `[]` is common? It looks weird to me. Or this example aims to demonstrate how to load a on-disk numpy array. If yes, it's better to use a 2-D numpy array instead of `np.arange`.

File: tests/python/pytorch/graphbolt/impl/test_basic_feature_store.py [hunk 13]
```diff
@@ -16,56 +16,60 @@ def test_basic_feature_store_homo():
...
 
 def test_basic_feature_store_errors():
     a = torch.tensor([3, 2, 1])
-    b = torch.tensor([2, 5, 3])
+    b = torch.tensor([[1, 2, 4], [2, 5, 3]])
 
     features = {}
-    features[("node", "paper", "a")] = gb.TorchBasedFeature(a)
+    # Test error when key does not exist.
+    with pytest.raises(AssertionError):
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/impl/test_basic_feature_store.py:71]: the assertion error is thrown by dimension check or key existence check? please capture the error message as well. Reference:
```
with pytest.raises(
            RuntimeError,
            match=rf"Dataset {dataset_name} is not available.*",
        ):
```

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 14]
```diff
@@ -23,37 +23,45 @@ def __init__(self, torch_feature: torch.Tensor):
...
-        tensor([0, 1, 2, 3, 4])
-        >>> feature_store.read(torch.tensor([0, 1, 2]))
-        tensor([0, 1, 2])
-        >>> feature_store.update(torch.ones(3, dtype=torch.long),
-        ... torch.tensor([0, 1, 2]))
-        >>> feature_store.read(torch.tensor([0, 1, 2, 3]))
-        tensor([1, 1, 1, 3])
+        tensor([[0, 1, 2, 3, 4],
+                [5, 6, 7, 8, 9]])
+        >>> feature_store.read(torch.tensor([0]))
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:37]: feature_store -> feature for all.

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 15]
```diff
@@ -23,37 +23,45 @@ def __init__(self, torch_feature: torch.Tensor):
...
+        >>> torch_feat = torch.from_numpy(np.load("/tmp/arr.npy", mmap_mode="r+"))
+        >>> feature_store = gb.TorchBasedFeature(torch_feat)
         >>> feature_store.read()
-        tensor([0, 1, 2, 3, 4])
-        >>> feature_store.read(torch.tensor([0, 1, 2]))
-        tensor([0, 1, 2])
+        tensor([[1, 2],
+                [3, 4]])
+        >>> feature_store.read(torch.tensor([0]))
+        tensor([[0, 1, 2, 3, 4]])
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:54]:
```suggestion
        tensor([[1, 2]])
```

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/impl/torch_based_feature_store.py [hunk 16]
```diff
@@ -23,37 +23,45 @@ def __init__(self, torch_feature: torch.Tensor):
...
             The torch feature.
+            Note that the dimension of the tensor should be greater than 1.
 
         Examples
         --------
         >>> import torch
-        >>> torch_feat = torch.arange(0, 5)
-        >>> feature_store = TorchBasedFeature(torch_feat)
+        >>> from dgl import graphbolt as gb
+        >>> torch_feat = torch.arange(10).reshape(2, -1).to("cuda")
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/torch_based_feature_store.py:32]: why `to("cuda")`?

**Review by Rhett-Ying — APPROVED** [review]
